adcc/solver/LanczosIterator.py

```diff
@@ -0,0 +1,226 @@
+##
+## adcc is free software: you can redistribute it and/or modify
+## it under the terms of the GNU General Public License as published
+## by the Free Software Foundation, either version 3 of the License, or
+## (at your option) any later version.
+##
+## adcc is distributed in the hope that it will be useful,
+## but WITHOUT ANY WARRANTY; without even the implied warranty of
+## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+## GNU General Public License for more details.
+##
+## You should have received a copy of the GNU General Public License
+## along with adcc. If not, see <http://www.gnu.org/licenses/>.
+##
+## ---------------------------------------------------------------------
+import copy
+import numpy as np
+
+from adcc import evaluate, lincomb
+from adcc.timings import Timer
+from adcc.AdcMatrix import AdcMatrixlike
+from adcc.AmplitudeVector import AmplitudeVector
+
+from .orthogonaliser import GramSchmidtOrthogonaliser
+
+
+class LanczosIterator:
+    def __init__(self, matrix, guesses, ritz_vectors=None, ritz_values=None,
+                 ritz_overlaps=None, explicit_symmetrisation=None):
+        if not isinstance(matrix, AdcMatrixlike):
+            raise TypeError("matrix is not of type AdcMatrixlike")
+        n_problem = matrix.shape[1]   # Problem size
+
+        if not isinstance(guesses, list):
+            guesses = [guesses]
+        for guess in guesses:
+            if not isinstance(guess, AmplitudeVector):
+                raise TypeError("One of the guesses is not an AmplitudeVector")
+        n_block = len(guesses)  # Lanczos block size
+
+        # For thick restarts, defaults to no restart
+        if ritz_values is None:
+            n_restart = 0
+            ritz_vectors = []  # Y
+            ritz_overlaps = np.empty((0, n_block))  # Sigma
+            ritz_values = np.empty((0, ))  # Theta
+        else:
+            n_restart = len(ritz_values)
+            if len(ritz_vectors) != n_restart or \
+               ritz_overlaps.shape != (n_restart, n_block):
+                raise ValueError("Restart vector shape does not agree "
+                                 "with problem.")
+
+        self.matrix = matrix
+        self.ritz_values = ritz_values
+        self.ritz_vectors = ritz_vectors
+        self.ritz_overlaps = ritz_overlaps
+        self.n_problem = n_problem
+        self.n_block = n_block
+        self.n_restart = n_restart
+        self.ortho = GramSchmidtOrthogonaliser(explicit_symmetrisation)
+        self.explicit_symmetrisation = explicit_symmetrisation
+        self.timer = Timer()  # TODO More fine-grained timings
+
+        # To be initialised by __iter__
+        self.lanczos_subspace = []
+        self.alphas = []  # Diagonal matrix block of subspace matrix
+        self.betas = []   # Side-diagonal matrix blocks of subspace matrix
+        self.residual = guesses
+        self.n_iter = 0
+        self.n_applies = 0
+
+    def __iter__(self):
```

Comment:
Where are these functions actually called? 😄 
I'm a bit overwhelmed by the Python magic probably...

Reply:
Well basically this is an iterator that "iterates" through the Lanczos subspace. They are implicitly called in the loop "for subspace in iterator" in the lanczos.py file and setup and advance the state.

Reply:
That's cool 👍